web/packages/teleport/src/Account/ChangePasswordWizard/ChangePasswordWizard.tsx

```diff
@@ -0,0 +1,338 @@
+import { useAsync } from 'shared/hooks/useAsync';
+import { Auth2faType } from 'shared/services';
+
+import { ChangePasswordReq } from 'teleport/services/auth';
+import auth, { MfaChallengeScope } from 'teleport/services/auth/auth';
+import { MfaDevice } from 'teleport/services/mfa';
+import { DialogHeader } from '../DialogHeader';
+
+export interface ChangePasswordWizardProps {
+  /** MFA type setting, as configured in the cluster's configuration. */
+  auth2faType: Auth2faType;
+  /** Determines whether the cluster allows passwordless login. */
+  passwordlessEnabled: boolean;
+  /** A list of available authentication devices. */
+  devices: MfaDevice[];
+  onClose(): void;
+  onSuccess(): void;
+}
+
+export function ChangePasswordWizard({
+  auth2faType,
+  passwordlessEnabled,
+  devices,
+  onClose,
+  onSuccess,
+}: ChangePasswordWizardProps) {
+  const reauthOptions = createReauthOptions(
+    auth2faType,
+    passwordlessEnabled,
+    devices
+  );
+  const [reauthMethod, setReauthMethod] = useState<ReauthenticationMethod>(
+    reauthOptions[0]?.value
+  );
+  const [credential, setCredential] = useState<Credential | undefined>();
+  const reauthRequired = reauthOptions.length > 0;
+
+  return (
+    <Dialog
+      open={true}
+      disableEscapeKeyDown={false}
+      dialogCss={() => ({ width: '650px' })}
+      onClose={onClose}
+    >
+      <StepSlider
+        flows={wizardFlows}
+        currFlow={
+          reauthRequired ? 'withReauthentication' : 'withoutReauthentication'
+        }
+        // Step properties
+        reauthOptions={reauthOptions}
+        reauthMethod={reauthMethod}
+        credential={credential}
+        onReauthMethodChange={setReauthMethod}
+        onAuthenticated={setCredential}
+        onClose={onClose}
+        onSuccess={onSuccess}
+      />
+    </Dialog>
+  );
+}
+
+type ReauthenticationMethod = 'passwordless' | 'mfaDevice' | 'otp';
+type ReauthenticationOption = {
+  value: ReauthenticationMethod;
+  label: string;
+};
+
+export function createReauthOptions(
+  auth2faType: Auth2faType,
+  passwordlessEnabled: boolean,
+  devices: MfaDevice[]
+) {
+  const options: ReauthenticationOption[] = [];
+
+  if (passwordlessEnabled) {
+    options.push({ value: 'passwordless', label: 'Passkey' });
+  }
+
+  const mfaEnabled = auth2faType === 'on' || auth2faType === 'optional';
+
+  if (auth2faType === 'webauthn' || mfaEnabled) {
+    options.push({ value: 'mfaDevice', label: 'MFA Device' });
+  }
+
+  if (auth2faType === 'otp' || mfaEnabled) {
+    options.push({ value: 'otp', label: 'Authenticator App' });
+  }
+
+  const allowedMethods = {};
+  for (const d of devices) {
+    allowedMethods[reauthMethodForDevice(d)] = true;
+  }
+
+  return options.filter(o => o.value in allowedMethods);
+}
+
+function reauthMethodForDevice(d: MfaDevice): ReauthenticationMethod {
+  if (d.usage === 'passwordless') return 'passwordless';
+  return d.type === 'totp' ? 'otp' : 'mfaDevice';
+}
```

Comment:
I wanted to avoid using `find()`, since technically, this is a collection of unknown size. But I agree this fragment is a bit cryptic, I'll change it.